useCountdown Hook

ndresx opened the pull request:
Full rewrite of the internals: all countdown logic now lives in a framework-agnostic `CountdownJs` engine, with React as a thin adapter over a new public `useCountdown` hook. `<Countdown />` is built on that hook, so component and hook share one lifecycle.

Ships dual ESM + CJS with bundled types, `sideEffects: false`, and targets React 18+. Changes are relative to 2.3.6.

## Breaking
- Requires **React 18+** (adapter uses `useSyncExternalStore`).
- Ref API moved: `ref.current.getApi()` → `ref.current.api`.
- Removed `className`, `children`, the legacy `count` prop, and `prop-types`.
- `onComplete` is now `(timeDelta, completedOnStart)`.

## Added
- Public **`useCountdown`** hook (same render props as the component).
- Subpath entries: `react-countdown/component` and `react-countdown/hook`.
- **`resetKey`** (restart without remount) and **`freezeProps`** (opt out of prop tracking).
- **`CountdownStatus`** enum export + **`getStatus()`** + **`refresh()`** API methods.

## Toolchain
- Updated all dev deps to latest: Jest 30, TypeScript 6, Cypress 15, Rollup 4.62, plugin bumps.

ndresx committed:
Initial commit with useCountdown hook

ndresx committed:
Refine hook, implement tests, update docs

ndresx committed:
Rebase with master, update dependencies

ndresx committed:
Update React peer dependencies for hook support

ndresx committed:
Re-add index

ndresx committed:
Merge remote-tracking branch 'origin/master' into use-countdown-hook

ndresx committed:
Update packages

ndresx committed:
Migrate to ESLint

ndresx committed:
Update ESLint rules

SalahAdDin commented:
@ndresx Do you have any plan to release this?

I'm having problems to use the `start()` function as i explained before in other issue, now i'm trying to get the `api()` from the render but i can't:
```
const Timer = ({ hours, minutes, seconds, completed, props }) => {
  const { className, renderer } = props;
  let time = completed ? `Is completed!` : `${hours}:${minutes}:${seconds}`;
  console.log('====================================');
  console.log(renderer.api.start());
  console.log('====================================');
  return (
    <StepLabel
      alternativeLabel
      StepIconComponent={StepIcon}
      StepIconProps={{
        classes: {
          root: className
        },  
        icon: time
      }}
    />
  );
};
```
`render.api` is undefined, getting `api` from `props` gives me the same result, i don't know what is the problem.

Thanks.

ndresx commented:
@SalahAdDin I have, but I am still working and tweaking some things. I have posted a possible solution into the other issue thread, please take a look at the provided code there; I hope it helps!

ndresx committed:
Merge branch 'master' into use-countdown-hook

ndresx committed:
Remove span and classname prop

SalahAdDin commented:
:+1:

ndresx committed:
Update tests, dependencies and other stuff

ndresx committed:
Add raf prop (temp)

ndresx committed:
Remove raf prop (experimental)

ndresx committed:
Remove/deprecated children prop support

ndresx committed:
Update examples, lint commands, node modules

ndresx committed:
Merge remote-tracking branch 'origin/master' into use-countdown-hook

# Conflicts:
#	package.json
#	src/Countdown.test.tsx
#	src/Countdown.tsx

ndresx committed:
Update travis cfg

ndresx committed:
Update examples build

kishoretankhupp commented:
@ndresx Any plan on this to merged in master so we can use it in react-native as well.

ndresx commented:
@kishoretankhupp Yes, but it, unfortunately, might still take a bit until then. I hope I can wrap it up soon though!

SalahAdDin commented:
@ndresx We are still waiting for this.

ndresx committed:
Merge remote-tracking branch 'origin/master' into use-countdown-hook

ndresx committed:
Update with master, update of all dependencies

ndresx committed:
Update lint steps

sahandsn commented:
any idea when it will merge?

SalahAdDin commented:
> any idea when it will merge?

It seems this repo is not maintained anymore.

sahandsn commented:
any suggestion about replacements? i didn't fine any package that has reference for now...

SalahAdDin commented:
> Any suggestions about replacements? I didn't find any package that has a reference for now...

Neither me, maybe a fork?

ndresx committed:
Remove PropTypes and update ESLint configuration for React component

ndresx committed:
Refactor countdown component and hook structure

- Updated import paths for Countdown component and useCountdown hook to reflect new directory structure.
- Modified CountdownApiExampleState to use the new component import.
- Changed CountdownHookBasicUsage and CountdownHookCompletionist to utilize UseCountdownProps.
- Updated package.json to specify new module paths and added exports for component and hook.
- Removed rollup.config.js and replaced it with rollup.config.mjs for improved module bundling.
- Added tests for component and hook entry points to ensure correct exports.
- Created separate component.ts and hook.ts files to re-export Countdown and useCountdown respectively.

ndresx committed:
Enhance countdown functionality and update dependencies

- Improved countdown behavior by allowing the `pure` prop to control re-renders.
- Updated peer dependencies to require React 18.
- Refactored Countdown and CountdownJs classes to simplify state management.
- Added unit tests for `calcOffsetStartTimestamp` method in CountdownJs.

ndresx committed:
Refactor testing setup by removing unused react-hooks library and updating imports

ndresx committed:
Implement structural updates and optimizations across multiple modules

ndresx committed:
Refactor code structure for improved readability and maintainability

ndresx committed:
Refactor Countdown component to expose control API via ref and update tests accordingly

ndresx committed:
Refactor Countdown component to utilize useCountdown hook and streamline API exposure

ndresx committed:
Refactor Countdown component to replace `pure` prop with `freezeProps` for improved prop handling

ndresx committed:
Refactor Countdown component to introduce resetKey prop for restarting countdown without remounting

ndresx committed:
Update README to require React 18 or higher

ndresx committed:
Enhance CI workflow to support React 18 and 19 matrix testing

ndresx committed:
Update Node.js version in CI workflow to 24.x

ndresx committed:
Update CI workflow and dependencies for React 19 compatibility

ndresx committed:
Update README to clarify server-side rendering hydration mismatch and provide solutions

ndresx committed:
Update dependencies to React 19 and add CountdownRef component for imperative control

ndresx committed:
Add Stopwatch example component and related tests for overtime functionality

ndresx committed:
Add Stopwatch component and update related examples and tests for improved functionality

ndresx committed:
Refactor Countdown component and tests to expose CountdownStatus enum and enhance API with getStatus method

ndresx committed:
Update README and tests to enhance documentation

ndresx committed:
Refactor code structure for improved readability and maintainability, migration to pnpm

ndresx committed:
Update README and Countdown component documentation for clarity and accuracy

ndresx committed:
Enhance README and tests to clarify new onTick behavior and improve CountdownJs functionality

ndresx committed:
Add tests for controlled Countdown behavior and prevent paused value drift

ndresx committed:
Add prepublishOnly script to ensure build before publishing

ndresx committed:
Add CHANGELOG.md for version 3.0.0 with breaking changes and new features

ndresx committed:
Refactor CodeQL workflow and update README for clarity and changelog reference

ndresx committed:
Enhance CI workflow to support React version matrix for testing and add conditional steps for React 19

ndresx committed:
Update README and CHANGELOG for v3 beta release information

ndresx committed:
Refactor

ndresx committed:
Update README to clarify `useCountdown` Hook props and `total` value description

ndresx committed:
feat: add setState test for CountdownJs to ensure state merging and listener notification

chore: update Jest snapshot comments to point to official documentation

chore: update Jest snapshot comments to point to official documentation

refactor: change TypeScript module resolution to 'bundler' and specify Jest types explicitly

SalahAdDin commented:
Too big to make a review.

ndresx committed:
docs: update README to include mention of useCountdown Hook

ndresx committed:
feat: add caching and installation steps for Cypress binary in CI workflow

ndresx committed:
chore: update actions to v5 in workflows for consistency and improvements

ndresx committed:
feat: add dry run script for package packing

ndresx committed:
feat: add refresh method to Countdown API and corresponding tests
